Cubical/Algebra/Ring/Base.agda

```diff
@@ -132,6 +132,7 @@ makeRing 0r 1r _+_ _·_ -_ is-setR assoc +-rid +-rinv +-comm ·-assoc ·-rid ·-
        (makeIsRing is-setR assoc +-rid +-rinv +-comm
                    ·-assoc ·-rid ·-lid ·-rdist-+ ·-ldist-+ )
 
+-- TODO: why are the Ring structures split and organized like this?
```

Comment:
@felixwellen  @mzeuner Do you have any clue about this? It seems like a bad idea to me as Agda will have trouble to figure out the ring structures

Reply:
Maybe @ecavallo knows?

Reply:
Can you explain the problem a bit more? I don't get what problems will be hard to solve for Agda and which aspect of the definition should be the cause. 
If the type of 'dist' is a problem - that could be something I did a long time ago and would do different now.

Reply:
I would like to change this to:
```
record IsRingHom {R : Ring ℓ} {S : Ring ℓ'} (f : R .fst → S .fst) : Type (ℓ-max ℓ ℓ')
```

I think if one keeps the ring structure separate from the type then Agda can sometimes get confused and not automatically infer it, forcing us to give it explicitly. I might be wrong, but I think this could happen more often when we consider homomorphisms between complex rings (e.g. polynomial rings over localized rings, etc.)

Reply:
Oh - I thought that style of defining structures was the result of a discussion about things like that. In the beginning, it was different and now all algebraic structures (as far as I know) are defined like this.

Reply:
I think it was my decision. The idea is just that it matches the format of a dependent path type (and consequently the format of a displayed univalent relation).

I don't follow how the change you're suggesting would help with inference (but then there's a lot I don't understand about Agda's inference). Note that in the current definition the structures are explicit arguments. I would expect to have to mention them explicitly as often as one needs to provide the endpoints explicitly in functions manipulating dependent paths.

Reply:
Hmm, OK... Can @ecavallo help me remember why we wanted it like this?  Note that I'm really only asking about what the parameters to homomorphisms should be

Reply:
Ah right, I guess the rings should be explicit in my proposal. No matter what, wouldn't this be slightly nicer:

```
record IsRingHom {A : Type ℓ} {B : Type ℓ'} (R : RingStr A) (S : RingStr B) (f : A → B) 
```

?

Reply:
I guess the order `{A : Type ℓ} {B : Type ℓ'} (f : A → B) (R : RingStr A) (S : RingStr B)` would actually be the closest to dependent paths. Anyway, I don't really prefer one over the other, but switching to a different convention would involve rewriting a lot of code!

Reply:
The bigger problem I had with the current organization, by the way, is that because `IsRingEquiv` is just a wrapper around `IsRingHom` (instead of `isRingEquiv` being its own record), Agda often forgets all but the function part of the equivalence argument and needs to be reminded explicitly. I was trying to decide what would be best to do with this when I put the algebra stuff to the side and never came back...

Reply:
Yes, this is exactly the kind of problem I've been trying to explain... Similar problems made `CommRingHom` annoying to use in my other PR

Reply:
OK, but that seems like a problem with the definition of `CommRingHom` which can't be fixed by changing `IsRingHom`? Or am I misunderstanding?

Reply:
One option I considered was to parameterize records like `IsRingHom` by an arbitrary notion of structure from which a ring could be projection, sort of a "typeclass" approach. Then you could instantiate at rings or commrings, to get the benefits of working with a record that hangs on to all the structure information without having to write everything multiple times:
``
Ringable : (∀ {ℓ} → Type ℓ → Type ℓ) → Typeω
Ringable S = {ℓ : Level} {A : Type ℓ} → S A → RingStr A

RingableRing : Ringable RingStr
RingableRing = idfun _

record IsRingHomGen {Str : ∀ {ℓ} → Type ℓ → Type ℓ} (π : Ringable Str)
  {ℓ} {ℓ'} {A : Type ℓ} {B : Type ℓ'}
  (R : Str A) (f : A → B) (S : Str B)
  : Type (ℓ-max ℓ ℓ')
  where

  -- Shorter qualified names
  private
    module R = RingStr (π R)
    module S = RingStr (π S)

  field
    pres0 : f R.0r ≡ S.0r
    pres1 : f R.1r ≡ S.1r
    pres+ : (x y : A) → f (x R.+ y) ≡ f x S.+ f y
    pres· : (x y : A) → f (x R.· y) ≡ f x S.· f y
    pres- : (x : A) → f (R.- x) ≡ S.- (f x)

IsRingHom = IsRingHomGen RingableRing
IsCommRingHom = IsRingHomGen ...
``
One could presumably do the same thing with the function/equivalence argument. I didn't spend enough time experimenting with this to know it would be a good idea, though.

Reply:
How about we take this discussion to an issue?